### ISSUE TITLE
Feature: Cluster Autoscaling

### ISSUE THREAD

**Issue by mumoshu** [kubernetes-incubator/kube-aws#148]
a.k.a [cluster-autoscaler](https://github.com/kubernetes/contrib/tree/master/cluster-autoscaler) Integration

I'd suggest something like the below in cluster.yaml to enable autoscaling of worker nodes based on pending pods.

```
# an auto scaling group
# see #142
worker:
  autoScalingGroup:
    minSize: 1
    maxSize: 20

# or a spot fleet (see #112)

worker:
  spotFleet:
    minSize: 1
    maxSize: 20

# which auto-scaling is enabled like:

experimental:
  workerClusterAutoscaler:
    enabled: true

# or

worker:
  clusterAutoScaler
    enabled: true

# or

worker:
  experimental:
    clusterAutoScaler:
      enabled: true
```

After thinking for a while about how we'd like [cluster-autoscaler](https://github.com/kubernetes/contrib/tree/master/cluster-autoscaler) to be deployed and managed to achieve auto-scaling of worker nodes in a main cluster or a node pool created via kube-aws, it turns out to be a bit cumbersome to fully automate it. It won't be a matter of just deploying it via `kubectl` and forget, at least.

## TL;DR;

How could we converge cluster-autoscaler to the desired state, without persisting the state outside of cluster?

## General questions

  * [x] To which node should we deploy cluster-autoscaler, controllers or workers?
     * Deploying it to worker nodes means it eats some resource we originally reserved for our application
     * cluster-autoscaler [is scheduled to master nodes on GCE, GKE](https://github.com/kubernetes/contrib/tree/master/cluster-autoscaler#deployment)
     * #150 is required if we'd like to deploy cluster-autoscaler to controller nodes
  * [x] Should we annotate cluster-autoscaler to be a "CriticalPod"?
     * We'd want cluster-autoscaler to survive resource shortage or we may end up unable to auto-scale nodes when in resource shortage(!)
     * If we'd like to annotate like that, rescheduler must be supported beforehand (#118)
     * [It is recommended to be a critical pod in cluster-autoscaler's doc](https://github.com/kubernetes/contrib/tree/master/cluster-autoscaler#deployment)

## kube-aws specific questions

  * How should we deploy cluster-autoscaler?
     * Via `kubectl apply -f <yaml file templated with vars like asg min/max size and asg name>`in cloud-config of worker nodes?
        * How could we let cluster-autoscaler stop managing specific ASG if it becomes unneeded?
            * How could we let cluster-autoscaler stop managing already deleted ASGs?
            * If we enabled auto-scaling on a ASG initially and then disabled afterwards, how could we let cluster-autoscaler stop managing the ASG?
  * How should we update cluster-autoscaler when an ASG's min/max size changed?
     * cluster-autoscaler is run like:
        ```
        cluster-autoscaler \
          --nodes <min size>:<max size>:<asg1> \
          --nodes <min size>:<max size>:<asg2> \
          ...
        ```
         to limit min/max num of nodes in an ASG.
     * If an ASG's min/max size is changed by `kube-aws update`, how should we update cluster-autoscaler accordingly?
        * `kubeclt apply -f something` to update cluster-autoscaler's args
           * Or configmap+inotifywait to watch for a configuration change?
        * How could we know which ASGs and spot fleets are linked to worker nodes?
           * Filter ASGs with the KubernetesCluster tag?
               * Doesn't work for Spot Fleet since it doesn't support tagging
                  * Filter CFN stacks with the KubernetesCluster tag and then find spot fleets managed by them? A bit cumbersome but would work.
        * If we don't do this, cluster-autoscaler won't scale (more than the max size|less than the min size) set before `kube-aws update` until we manually update it.

**Comment by mumoshu** [comment]
I'd rather implement a control loop, which is a cluster-side singleton, to converge cluster-autoscaler 's state(i.e. a sequence of `--nodes <min>:<max>:<asg name>`) to its desired state(i.e. parameters of ASGs with auto-scaling enabled at a point of time).

Ideally it should be self-hosted like kube-system components but I'm going to start with a POC implemented as a simple shell script.

**Comment by mumoshu** [comment]
https://github.com/kubernetes/contrib/pull/2226 is in progress to allow cluster-autoscaler to be automatically updated according to the state of a kube-aws cluster.

**Comment by c-knowles** [comment]
Helm Chart for [cluster-autoscaler](https://github.com/kubernetes/charts/tree/master/stable/aws-cluster-autoscaler also ) now also available.

**Comment by mumoshu** [comment]
I'm working on https://github.com/kubernetes/autoscaler/pull/11 to allow CA to be integrated with kops, kube-aws, etc. easier than ever.

**Comment by mumoshu** [comment]
https://github.com/kubernetes/autoscaler/pull/11 is now merged.
I'm now working on #629

**Comment by mumoshu** [comment]
#629 is merged. Closing this as resolved. Please submit new issue(s) if you find any problems in the revised CA support :bow:

**Comment by mumoshu** [comment]
As the milestone assigned to this issue says, the revised CA support is included since v0.9.7-rc.1 but please use v0.9.7-rc.2 so that you won't be affected by the terrible bug I made 😝
